File: IKSimulator/servo.py
import os

ELBOW_PIN = 1
BASE_PIN = 0

class Servo:
    def __init__(self, pin):
        self.pin = pin
        if pin == ELBOW_PIN:
            self.invert = False
            self.minDeg = 47
            self.maxDeg = 135
            self.minPulse = 1000
            self.maxPulse = 2100
        elif pin == BASE_PIN:
            self.invert = True
            print("Base")
            self.minDeg = 0
            self.maxDeg = 90
            self.minPulse = 930 #90 Deg
            self.maxPulse = 2200

        self.pulse(1500)

    def stop(self):
        self.set(0)

    def pulse(self, pulse):
        command = "echo "+str(self.pin)+"="+str(int(pulse/10))+" > /dev/servoblaster"
        print(command)
        os.system(command)

    def set(self, deg):
        self.pulsePerDeg = float(self.maxPulse - self.minPulse) / (self.maxDeg - self.minDeg)
        if self.invert: #Base Servo!
            pulse = self.maxPulse-(deg*self.pulsePerDeg)
        else:           #Elbow Servo!
            pulse = self.minPulse+(deg*self.pulsePerDeg)

        self.pulse(pulse)

File: RPi_Server_CPP/servo.cpp
#include <stdio.h>
#include <iostream>

class Servo{
	static const int ELBOW_PIN = 0;
	static const int BASE_PIN = 1;
	FILE * servoDev;
	bool invert;
	int servoPin, minDeg, maxDeg, minPulse, maxPulse;
	void pulse(int pulse){
		fprintf(servoDev, "%d=%d\n", servoPin, (int)(pulse/10));
		fflush(servoDev);
	}
public:
	void init(){
		servoDev = fopen("/dev/servoblaster", "w");
		if(servoPin == ELBOW_PIN){
			invert = false;
			minDeg = 35;
			maxDeg = 145;
			minPulse = 900;
			maxPulse = 2300;
		}
		else if(servoPin == BASE_PIN){
			invert = true;
			minDeg = -5;
			maxDeg = 105;
			minPulse = 900;
			maxPulse = 2300;
		}

		pulse((maxPulse + minPulse) / 2);
	}
	Servo(){
		servoPin = 0;
		init();
	}
	Servo(int pin){
		servoPin = pin;
		init();
	}
	bool set(double deg){
		std::cout << servoPin << ": " << deg << std::endl;
		if(deg < minDeg || deg > maxDeg)
			return false;
		double pulsePerDeg = (maxPulse - minPulse) / (maxDeg - minDeg);
		if(invert){
			pulse(maxPulse - (deg * pulsePerDeg));
		} else {
			pulse(minPulse + (deg * pulsePerDeg));
		}
		return true;
	}
	void off(){
		set(0);
	}
	~Servo(){
		off();
		fclose(servoDev);
	}
};
